// Copyright 2020 Google LLC. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
syntax = "proto3";

package tfx.orchestration;

import "google/protobuf/descriptor.proto";
import "ml_metadata/proto/metadata_store.proto";

// PlaceholderExpression is used in DSL IR executor specs to represent
// information that is not available at the authoring time. It will be rendered
// at execution time to provide the needed information.
message PlaceholderExpression {
  oneof expression_type {
    // ml_metadata.Value currently does not support boolean type.
    // Boolean values are stored as string literals "true" or "false".
    ml_metadata.Value value = 1;
    Placeholder placeholder = 2;
    PlaceholderExpressionOperator operator = 3;
  }
}

// Placeholder operator helps extract or render placeholders.
// TODO(b/167431029): Add custom options to annotate operator types.
message PlaceholderExpressionOperator {
  oneof operator_type {
    ArtifactUriOperator artifact_uri_op = 1;
    ArtifactValueOperator artifact_value_op = 2;
    ConcatOperator concat_op = 3;
    IndexOperator index_op = 4;
    ProtoOperator proto_op = 5;
  }
}

// ArtifactUriOperator extracts the Artifact URI from a placeholder expression.
// ArtifactUriOperator: Artifact -> String
message ArtifactUriOperator {
  // Required. It must evaluate to an INPUT_ARTIFACT or OUTPUT_ARTIFACT type
  // placeholder.
  PlaceholderExpression expression = 1;

  // Optional. If provided, this operator returns the URI for a specific split.
  // It's needed if one artifact, e.g., Examples Artifact, has multiple splits.
  string split = 2;
}

// ArtifactValueOperator evaluates to the value of primitive-type artifacts
// from a placeholder expression, by reading the contents at the Artifact's URI.
// ArtifactValueOperator: Artifact -> String
message ArtifactValueOperator {
  // Required. It must evaluate to a primitive-type artifact.
  PlaceholderExpression expression = 1;
}

// ConcatOperator concatenates multiple placeholder expressions.
// ConcatOperator: List[String] -> String
message ConcatOperator {
  // Required. The placeholders to be concatenated.
  repeated PlaceholderExpression expressions = 1;
}

// IndexOperator extracts an element at the index if the PlaceholderExpression
// evaluates to something repeated. For example, an INPUT_ARTIFACT or
// OUTPUT_ARTIFACT placeholder may contain multiple artifacts.
// IndexOperator: List[Any] -> Any
message IndexOperator {
  // Required. The placeholder expression must evaluate to something repeated.
  // Otherwise a runtime error will be thrown when evaluating this operator.
  PlaceholderExpression expression = 1;

  // Required. The element index to be accessed.
  int64 index = 2;
}

// ProtoOperator helps access and serialize the information in
// PlaceholderExpression if it evaluates to a proto.
// ProtoOperator: Proto -> String
message ProtoOperator {
  message ProtoSchema {
    // Message type for the proto. It must exist in the file_descriptors below.
    string message_type = 1;
    // Definitions of all dependencies of the above message type.
    google.protobuf.FileDescriptorSet file_descriptors = 2;
  }

  // Required. An expression that evaluates to a proto type.
  // Typically, this is an EXEC_PROPERTY type placeholder.
  PlaceholderExpression expression = 1;

  // Required. Proto schema makes proto-valued placeholder expression
  // self-describing.
  ProtoSchema proto_schema = 2;

  // Optional. Access a field inside the proto, using proto notations.
  // If provided, the value at the proto_field_path is a non-message type,
  // it will be rendered directly by converting to a string.
  // If this field is not provided, or the proto_field_path points to a
  // sub-message, that value will be rendered to a string using text_format.
  //
  // For example: proto_field_path=["[0]", ".a", ".b", "['c']"] is equivalent to
  // accessing p[0].a.b['c'], where p is the proto evaluated from the above
  // expression.
  repeated string proto_field_path = 3;

  // Supported schemes to serialize a proto.
  enum SerializationFormat {
    DEFAULT = 0;
    TEXT_FORMAT = 1;
    JSON = 2;
    BINARY = 3;
  }

  // Optional. The scheme that the proto is serialized to.
  SerializationFormat serialization_format = 4;
}


// Placeholder is an inseparable unit in a placeholder expression that
// pipeline information will be injected.
message Placeholder {
  // A Placeholder can represent an input artifact, an output artifact,
  // or an execution property.
  enum Type {
    // TODO(b/172065067): Clean up enum to reflect structure of the SDK objects.
    INPUT_ARTIFACT = 0;
    OUTPUT_ARTIFACT = 1;
    EXEC_PROPERTY = 2;
    RUNTIME_INFO = 3;
    EXEC_INVOCATION = 4;
  }
  // Required. Type of the placeholder.
  Type type = 1;

  // Required. Key to the placeholder.
  string key = 2;
}
